### DISCUSSION TITLE
What is the difference between these typescript types and which ones should I use?

### DISCUSSION THREAD

**Discussion by Axedyson** [storybookjs/storybook#16650] [Help] [answered]
When I ran `npx sb init` it set up storybook automatically in my `typescript, next.js and react` project. But I noticed that the typescript types for the default export and the template code are different from what is shown in the documentation examples:

Here is my story as an example:
`Button.stories.tsx`
```TypeScript
import { Button } from "./Button";
import type { ComponentMeta, ComponentStory } from "@storybook/react";

export default {
  title: "Button",
  component: Button,
  args: {
    label: "Button",
  },
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
```
As you can see I'm currently using the type `ComponentMeta<typeof Button>` for my Button story.
And at the template level I'm using `ComponentStory<typeof Button>`. But in the docs the following types are used:
```TypeScript
import { Button } from "./Button";
import { Meta, Story } from '@storybook/react';

export default {
  title: "Button",
  component: Button,
  args: {
    label: "Button",
  },
} as Meta;

const Template: Story<"something in here"> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
```
So what is the recommended type? Should I keep using `ComponentMeta` and `ComponentStory` or should I use `Meta` and `Story` instead?

**Comment by yannbf** [comment, marked as answer]
Great question @AndysonDK!
Storybook used to have only the `Story` and `Meta` types for all frameworks. In React, users had to export their component's prop types in order to provide typing information for Storybook:

```tsx
import type { Meta, Story } from '@storybook/react'
import { MyComponent, MyComponentProps } from './MyComponent' // import types from component
export default {
  component: MyComponent
} as Meta<MyComponentProps> // this gives autocompletion for args at the Meta level

const Template: Story<MyComponentProps> = (args) => <MyComponent {...args} /> // This gives autoCompletion for args at Story level
```

But realistically, it's not really common to export the types of a component's prop in React projects, and users were doing that just because of Storybook.

As a way to circumvent that, we released new types in Storybook 6.3 specifically for React called `ComponentStory` and `ComponentMeta`. They allow us to extract prop type information without requiring the user to export them. As a plus, this also works for inline prop types as well.

```tsx
import type { ComponentMeta, ComponentStory } from '@storybook/react'
import { MyComponent } from './MyComponent' // no need to import types anymore!
export default {
  component: MyComponent
} as Meta<typeof MyComponentProps> // types extracted from typeof MyComponentProps for args autocompletion

const Template: ComponentStory<typeof MyComponentProps> = (args) => <MyComponent {...args} /> // types extracted from typeof MyComponentProps for args autocompletion
```

TLDR: the `Component` versions of `Story` and `Meta` are special convenience types so that you don't have to export types from your component file. You can use either versions, it's really up to you!

**Reply by Inokcode** [reply]
```ts
import type { ComponentMeta, ComponentStory } from '@storybook/react'
import { MyComponent } from './MyComponent' // no need to import types anymore!
export default {
  component: MyComponent
} as Meta<typeof MyComponentProps> // types extracted from typeof MyComponentProps for args autocompletion

const Template: ComponentStory<typeof MyComponentProps> = (args) => <MyComponent {...args} /> // types extracted from typeof MyComponentProps for args autocompletion
```

**Reply by Inokcode** [reply]
above code,is this correct ?bcz without importing MyComponentProps and Meta how can we use it as a <typeof MyComponentProps>

**Reply by yannbf** [reply]
Hey @Inokcode the imports were missing in my example and in yours, I updated them. As mentioned in my comment:

> We released new types specifically for React called ComponentStory and ComponentMeta. They allow us to extract prop type information without requiring the user to export them. As a plus, this also works for inline prop types as well.

These types will _infer_ the props of your component, so it's not necessary to import them.

**Reply by radfahrer** [reply]
Maybe I'm missing something, but I don't actually see you using `ComponentMeta` is there a better example you can point me at?

**Reply by Axedyson** [reply]
> Maybe I'm missing something, but I don't actually see you using `ComponentMeta` is there a better example you can point me at?

I think it was a mistake, I think what he means by using `Meta<T>` is actually `ComponentMeta<T>`

**Reply by KevinGhadyani-Okta** [reply]
It'd be great if the TS descriptions were better:
> For the common case where a component's stories are simple components that receives args as props:
> `export default { ... } as ComponentMeta<typeof Button>;`

This message doesn't tell me why I'd wanna use one or the other or what this is.

When in TypeScript, I see something named `ComponentMeta` and try to use it only to immediately ask "why would they create this? How is it different than `Meta`?", and I found myself at this answer because the TS description didn't tell me.

**Comment by kamleshchandnani** [comment]
in what version it was introduced? I'm right now on sb v6.1 but the TS compiler throws `Module '"@storybook/react"' has no exported member 'ComponentMeta'.`

**Reply by shilman** [reply]
Introduced in 6.3.0

**Comment by ghost** [comment]
With polymorphic components, there is no  IntelliSense,
![image](https://user-images.githubusercontent.com/79634799/198870194-2033bb8c-9d57-416c-9d62-e1fdbaddb935.png)
![image](https://user-images.githubusercontent.com/79634799/198870209-de496093-8bda-4558-b2c3-16713016f381.png)
